File: docs/google-studio-handoff/complete-website/scripts/optimize-images.mjs
import fs from "node:fs/promises";
import path from "node:path";
import sharp from "sharp";

const root = process.cwd();
const assetsRoot = path.join(root, "assets");
const manifest = {};

async function walk(directory) {
  const entries = await fs.readdir(directory, { withFileTypes: true });
  const files = [];
  for (const entry of entries) {
    const fullPath = path.join(directory, entry.name);
    if (entry.isDirectory()) files.push(...(await walk(fullPath)));
    else files.push(fullPath);
  }
  return files;
}

const sources = (await walk(assetsRoot)).filter((file) => {
  const ext = path.extname(file).toLowerCase();
  return [".jpg", ".jpeg", ".png"].includes(ext) && !path.basename(file).includes(".optimized.") && path.basename(file) !== "favicon.png";
});

for (const source of sources) {
  try {
    const ext = path.extname(source).toLowerCase();
    const base = source.slice(0, -ext.length);
    const jpgPath = `${base}.jpg`;
    const webpPath = `${base}.webp`;
    const relativeSource = path.relative(root, source).replaceAll(path.sep, "/");
    if (relativeSource.startsWith("assets/factory/")) {
      try {
        await fs.access(webpPath);
        const metadata = await sharp(source).metadata();
        const dimensions = { width: metadata.width, height: metadata.height };
        manifest[relativeSource] = dimensions;
        manifest[path.relative(root, webpPath).replaceAll(path.sep, "/")] = dimensions;
        continue;
      } catch {}
    }
    const pipeline = sharp(source).autoOrient().resize({ width: 1800, withoutEnlargement: true });

    if (source === jpgPath) {
      const tempPath = `${base}.optimized.jpg`;
      await pipeline.clone().jpeg({ quality: 82, progressive: true, mozjpeg: true }).toFile(tempPath);
      await fs.rm(jpgPath, { force: true });
      await fs.rename(tempPath, jpgPath);
    } else {
      await pipeline.clone().jpeg({ quality: 82, progressive: true, mozjpeg: true }).toFile(jpgPath);
    }

    await pipeline.clone().webp({ quality: 78, effort: 6 }).toFile(webpPath);
    const metadata = await sharp(jpgPath).metadata();
    const dimensions = { width: metadata.width, height: metadata.height };
    manifest[path.relative(root, jpgPath).replaceAll(path.sep, "/")] = dimensions;
    manifest[path.relative(root, webpPath).replaceAll(path.sep, "/")] = dimensions;
  } catch (error) {
    console.warn(`Skipping unreadable image ${path.relative(root, source)}: ${error.message}`);
  }
}

await fs.writeFile(path.join(root, "data/image-manifest.json"), `${JSON.stringify(manifest, null, 2)}\n`);
console.log("Optimized images with sharp and generated JPEG/WebP variants.");

File: docs/google-studio-handoff/complete-website/scripts/optimize-images.py
import json
from pathlib import Path
from PIL import Image

ROOT = Path(__file__).resolve().parents[1]
ASSETS = ROOT / "assets"
MANIFEST = ROOT / "data" / "image-manifest.json"
manifest = {}

for path in ASSETS.rglob("*"):
    if path.suffix.lower() not in {".jpg", ".jpeg", ".png"}:
        continue
    if path.name == "favicon.png":
        continue

    with Image.open(path) as img:
        img = img.convert("RGB")
        max_width = 1800
        if img.width > max_width:
            ratio = max_width / img.width
            img = img.resize((max_width, int(img.height * ratio)), Image.Resampling.LANCZOS)

        if path.suffix.lower() in {".jpg", ".jpeg"}:
            jpg_path = path
        else:
            jpg_path = path.with_suffix(".jpg")

        img.save(jpg_path, "JPEG", quality=82, optimize=True, progressive=True)

        webp_path = path.with_suffix(".webp")
        img.save(webp_path, "WEBP", quality=78, method=6)

        manifest[jpg_path.relative_to(ROOT).as_posix()] = {"width": img.width, "height": img.height}
        manifest[webp_path.relative_to(ROOT).as_posix()] = {"width": img.width, "height": img.height}

MANIFEST.write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")

print("Optimized images and generated real WebP files.")
